File: src/lib/encoding.ts
export function encodeInp(input: string): string {
  return Buffer.from(input, "utf8").toString("base64");
}

/**
 * 构建教务系统登录的加密凭证。
 *
 * 算法：
 * 1. 将 account、password、空格分别做 base64 编码，用 "%%%" 拼接成 code
 * 2. 遍历 code 的前 55 个字符，根据 sxh 的数字决定在每字符后插入多少位 scode
 */
export function buildEncodedCredential(
  account: string,
  password: string,
  scodeSeed: string,
  sxh: string
): string {
  const accountEncoded = encodeInp(account);
  const passwordEncoded = encodeInp(password);
  const codeDogEncoded = encodeInp(" ");
  const code = `${accountEncoded}%%%${passwordEncoded}%%%${codeDogEncoded}`;

  let scode = scodeSeed;
  let encoded = "";

  for (let i = 0; i < code.length; i += 1) {
    if (i < 55) {
      const index = Number.parseInt(sxh.slice(i, i + 1), 10);
      const safeIndex = Number.isFinite(index) && index > 0 ? index : 0;
      encoded += `${code[i]}${scode.slice(0, safeIndex)}`;
      scode = scode.slice(safeIndex);
    } else {
      encoded += code.slice(i);
      break;
    }
  }

  return encoded;
}

File: main/encoding.py
"""生成旧版登录表单要求的凭据字段。"""

from __future__ import annotations

import base64


def _base64(value: str) -> str:
    return base64.b64encode(value.encode("utf-8")).decode("ascii")


def build_encoded_credential(
    account: str,
    password: str,
    scode_seed: str,
    sxh: str,
) -> str:
    """复现登录页 JavaScript 生成的 ``encoded`` 字段。"""

    code = f"{_base64(account)}%%%{_base64(password)}%%%{_base64(' ')}"
    scode = scode_seed
    output: list[str] = []

    for index, character in enumerate(code):
        if index >= 55:
            output.append(code[index:])
            break
        digit = sxh[index : index + 1]
        take = int(digit) if digit.isdigit() and digit != "0" else 0
        output.extend((character, scode[:take]))
        scode = scode[take:]

    return "".join(output)


__all__ = ["build_encoded_credential"]
